### PR TITLE
CSingleLock/CSharedLock/CExclusiveLock: replace with stl methods

### PR DESCRIPTION
**Pull request by lrusak**
This PR is a follow up to #20949. Reviewers felt that it would be best to just replace the classes completely.

The amount of changes is quite large. Most of it is find and replace.

`CSingleLock` -> `std::unique_lock<CCriticalSection>`
`CSharedLock` -> `std::shared_lock<CSharedSection>`
`CExclusiveLock` -> `std::unique_lock<CSharedSection>`

We also now use the `lock` and `unlock` methods directly.

My only concern here is that people may start using `std::mutex` and `std::shared_mutex` instead of our custom `CCriticalSection` and `CSharedSection` mutexes. That is something we may have to watch for and enforce.

clang-format did a number on [xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.cpp](https://github.com/xbmc/xbmc/compare/master...lrusak:locking-cleanup-PR-3?expand=1#diff-bad369057589a59119615d8f380bff839b51c2100f2e8b5fd92b24fb61ea3fc6) so it's probably better to look at the changes with whitespace changes off, click [here](https://github.com/xbmc/xbmc/pull/20980/files?w=1).

### CODE REVIEW

**Comment by garbear** [comment]
RetroPlayer, games, input and peripherals code looks as expected. Shouldn't we #include &lt;mutex> where std::unique_lock is introduced?

**Comment by phunkyfish** [comment]
Can you explain why `CCriticalSection` and `CSharedSection` need to be used instead of `std::mutex`? It would be nicer if it was all eventually STL.

**Comment by lrusak** [comment]
> RetroPlayer, games, input and peripherals code looks as expected. Shouldn't we #include <mutex> where std::unique_lock is introduced?

The build seems ok without it. I have a hard time adding includes blindly but will do so if required.

----

> Can you explain why `CCriticalSection` and `CSharedSection` need to be used instead of `std::mutex`? It would be nicer if it was all eventually STL.

Thanks for asking. The answer has three parts. @jimfcarroll can correct me if I get the explanation wrong.

1) using `CCriticalSection` inherits from `CountingLockable`, this is needed to allow counting the locks/unlocks and allows using the (anti-pattern) class `CSingleExit`. For those that don't know `CSingleExit` allows unlocking all the locks in a thread for a scoped period and restores the locks after the scope. I guess this behaviour is required for some of the original xbmc code.

2) On posix platforms `CCriticalSection` uses our custom `CRecursiveMutex` class. This `CRecursiveMutex` class is needed because we use the pthread attribute `PTHREAD_PRIO_INHERIT` which isn't possible to set using `std::recursive_mutex`. You can find an explanation for this attribute [here](https://linux.die.net/man/3/pthread_mutexattr_setprotocol). Basically this:
> When a thread is blocking higher priority threads because of owning one or more mutexes with the PTHREAD_PRIO_INHERIT protocol attribute, it shall execute at the higher of its priority or the priority of the highest priority thread waiting on any of the mutexes owned by this thread and initialized with this protocol.

3) This `PTHREAD_PRIO_INHERIT` is needed because we allow setting the thread priority. This uses the native underlying threading library for the platform and on posix this is pthreads. We allow setting some threads higher or lower or equal to the priority of the main app. This is questionable though really as on linux normal users cannot increase the priority beyond what the application is normally launched at (requires `CAP_SYS_NICE` or superuser or custom limits.conf). I believe the priority changing is limited to the linux though as we gate it with `#ifdef RLIMIT_NICE` which I believe is a linuxism.

So it's a large house of cards really. If we got rid of the ability to set priority we would also be able to remove some of the custom code (read `CRecursiveMutex`), however, our usage of `CSingleExit` basically requires us to use `CCriticalSection` and `CountingLockable`.

So in summary. It's not required to use `CCriticalSection` instead of `std::mutex` but you should be aware of the reasons. If you require an exit-able lock, or are in a thread with a higher priority it's required to use `CCriticalSection`. Same thing goes for `CSharedSection` instead of `std::shared_mutex`.

**Review by yol — APPROVED** [review]
Wayland changes look fine, even if I guess most of it could be replaced with std::lock_guard :-)

**Comment by garbear** [comment]
>> RetroPlayer, games, input and peripherals code looks as expected. Shouldn't we #include where std::unique_lock is introduced?

> The build seems ok without it. I have a hard time adding includes blindly but will do so if required.

The build succeeds because &lt;mutex.h> is pulled in from one of these files:

* ./threads/platform/RecursiveMutex.h:11:#include <mutex>
* ./utils/IPlatformLog.h:12:#include <mutex>
* ./addons/AddonManager.h:19:#include <mutex>

You can see how removing them from the include chain will break the build.

Every time I use a STL `std` or C stdlib function or member in a cpp I make sure that the corresponding header is pulled in by either its header (if appears it in the header) or the cpp itself. So it's not a blind process, but you can see how opaque the process is for someone who doesn't match all STL and stdlib members to their headers by memory alone.

So I would add &lt;mutex> to all cpp files where std::unique_lock replaces CSingleLock in this PR.

**Comment by garbear** [comment]
Just a general comment, but we could see breakage. CSingleLock opted authors into recursiveness. Replacing with `std::unique_lock` changes that assumption. Much of Kodi is surprisingly re-entrant, so if we wanted to avoid build breakage we'd replace everywhere with a recursive lock. But I'd argue to accept breakage so we can identify these places of re-entrancy.

**Comment by lrusak** [comment]
> Just a general comment, but we could see breakage. CSingleLock opted authors into recursiveness. Replacing with `std::unique_lock` changes that assumption. Much of Kodi is surprisingly re-entrant, so if we wanted to avoid build breakage we'd replace everywhere with a recursive lock. But I'd argue to accept breakage so we can identify these places of re-entrancy.

I'm not sure if you are implying that this change causes build breakage or that build breakage may occur in the future if people don't use `std::unique_lock<CCriticalSection>` in future code? :thinking:

**Comment by garbear** [comment]
I'm implying that this change may cause runtime breakage. A place that was using CSingleLock might accidentally have been re-entrant. CSingleLock is recursive, so no problem before, but now unique_lock isn't, so there may be a crash.

**Comment by lrusak** [comment]
> I'm implying that this change may cause runtime breakage. A place that was using CSingleLock might accidentally have been re-entrant. CSingleLock is recursive, so no problem before, but now unique_lock isn't, so there may be a crash.

Hmmm, I would think it would already be broken because of #20949

I'll see if I can do more testing but I would have expected our tests to fail but maybe there is an edge case somewhere.

**Comment by garbear** [comment]
> Hmmm, I would think it would already be broken because of https://github.com/xbmc/xbmc/pull/20949
> I'll see if I can do more testing but I would have expected our tests to fail but maybe there is an edge case somewhere.

I don't think we test for re-entrancy. I used the term "accidentally re-entrant", because it's an extreme edge case that is unintended to occur, but due to Kodi's pervasive re-entrancy nevertheless does occur in some weird edge case. Good luck finding it :) I propose we merge and look for these edge cases to show up in submitted crash logs.

**Review by garbear — APPROVED** [review]

**Comment by lrusak** [comment]
> > Hmmm, I would think it would already be broken because of #20949
> > I'll see if I can do more testing but I would have expected our tests to fail but maybe there is an edge case somewhere.
> 
> I don't think we test for re-entrancy. I used the term "accidentally re-entrant", because it's an extreme edge case that is unintended to occur, but due to Kodi's pervasive re-entrancy nevertheless does occur in some weird edge case. Good luck finding it :) I propose we merge and look for these edge cases to show up in submitted crash logs.

Let's find out 😉
